nameresolution/consul/consul.go

```diff
@@ -0,0 +1,327 @@
+
+			services[i], services[j] = services[j], services[i]
+		}
+
+		return services
+	}
+
+	svc := shuffle(services)[0]
+
+	// TODO - kubernetes behaviour?
+
+	addr := ""
+
+	if port, ok := svc.Service.Meta[cfg.DaprPortMetaKey]; ok {
+		if svc.Service.Address != "" {
+			addr = fmt.Sprintf("%s:%s", svc.Service.Address, port)
+		} else if svc.Node.Address != "" {
+			addr = fmt.Sprintf("%s:%s", svc.Node.Address, port)
+		} else {
+			return "", fmt.Errorf("no healthy services found with AppID:%s", req.ID)
+		}
+	} else {
+		return "", fmt.Errorf("target service AppID:%s found but DAPR_PORT missing from meta", req.ID)
+	}
+
+	return addr, nil
+}
+
+// getConfig configuration from metadata, defaults are best suited for self-hosted mode
+func getConfig(metadata nr.Metadata) (resolverConfig, error) {
+	var daprPort string
+	var ok bool
+	var err error
+	resolverCfg := resolverConfig{}
+
+	props := metadata.Properties
+
+	if daprPort, ok = props[nr.DaprPort]; !ok {
+		return resolverCfg, fmt.Errorf("nr metadata property missing: %s", nr.DaprPort)
+	}
+
+	cfg, err := parseConfig(metadata.Configuration)
+	if err != nil {
+		return resolverCfg, err
+	}
+
+	// set DaprPortMetaKey used for registring DaprPort and resolving from Consul
+	if cfg.DaprPortMetaKey == "" {
+		resolverCfg.DaprPortMetaKey = daprMeta
+	} else {
+		resolverCfg.DaprPortMetaKey = cfg.DaprPortMetaKey
+	}
+
+	resolverCfg.ClientConfig = getClientConfig(cfg)
+	if resolverCfg.Registration, err = getRegistrationConfig(cfg, props); err != nil {
+		return resolverCfg, err
+	}
+	resolverCfg.QueryOptions = getQueryOptionsConfig(cfg)
+
+	// if registering, set DaprPort in meta, needed for resolution
+	if resolverCfg.Registration != nil {
+		if resolverCfg.Registration.Meta == nil {
+			resolverCfg.Registration.Meta = map[string]string{}
+		}
+
+		resolverCfg.Registration.Meta[resolverCfg.DaprPortMetaKey] = daprPort
+	}
+
+	return resolverCfg, nil
+}
+
+func getClientConfig(cfg configSpec) *consul.Config {
+	// If no client config use library defaults
+	if cfg.ClientConfig != nil {
+		return cfg.ClientConfig
+	}
+
+	return consul.DefaultConfig()
+}
+
+func getRegistrationConfig(cfg configSpec, props map[string]string) (*consul.AgentServiceRegistration, error) {
+	// if advanced registration configured ignore other registration related configs
+	if cfg.AdvancedRegistration != nil {
+		return cfg.AdvancedRegistration, nil
+	} else if !cfg.SelfRegister {
+		return nil, nil
+	}
+
+	var appID string
+	var appPort string
+	var host string
+	var httpPort string
+	var ok bool
+
+	if appID, ok = props[nr.AppID]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppID)
+	}
+
+	if appPort, ok = props[nr.AppPort]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppPort)
+	}
+
+	if host, ok = props[nr.HostAddress]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.HostAddress)
+	}
+
+	if httpPort, ok = props[nr.DaprHTTPPort]; !ok {
```

Comment:
You might consider attempting to parse this as an unsigned integer for some level of validation.

Reply:
done